Adaptive pinned icons squircles (#1516)

ekager opened the pull request.

pocmo commented:
Forwarded the review to @mcomella who wrote the initial patch for adaptive shortcuts.

mcomella commented:
> so we'll have to figure out how to create the bitmap differently.

We do this in [`IconGenerator.generateAdaptiveLauncherIcon`](https://github.com/mozilla-mobile/focus-android/blob/01ec0e6cf8c260af62f20911b164c1feb0276517/app/src/main/java/org/mozilla/focus/shortcut/IconGenerator.kt#L65). Some approaches you could try:
- I could be getting the icon dimensions incorrect (though I took them from the AdaptiveIconDrawable source so probably not)
- I think converting the AdaptiveIconDrawable into the bitmap applies the transformation – we could just create the icon inside a regular Bitmap and pass that into `IconCompat`.

To be explicit, the transformation includes both rounding the corners and adjusting the padding, which is why I think the text size increases (it crops out the padding twice with your code).

ekager commented:
Okay I changed this to create a Bitmap with background and foreground and removed all Drawable conversions! Sorry I didn't delve into your comments the first time around! :)

mcomella commented on app/src/main/java/org/mozilla/focus/shortcut/IconGenerator.kt:
```diff
@@ -65,28 +59,37 @@ class IconGenerator {
          */
         @TargetApi(26)
-        private fun generateAdaptiveLauncherIcon(context: Context, url: String?): AdaptiveIconDrawable {
+        private fun generateAdaptiveLauncherIcon(context: Context, url: String?): Bitmap {
```
Most of this code is duplicated from `generateIcon`: can we merge the implementations? Duplication is bad because if we update one implementation, we often forget to update the others!

mcomella commented on app/src/main/java/org/mozilla/focus/shortcut/IconGenerator.kt:
```diff
@@ -40,60 +33,36 @@ class IconGenerator {
-        private fun generateIcon(context: Context, url: String?, iconSize: Int): Bitmap {
-            val bitmap = Bitmap.createBitmap(iconSize, iconSize, Bitmap.Config.ARGB_8888)
+        private fun drawCharacterOnBitmap(context: Context, url: String?, bitmap: Bitmap): Bitmap {
             val canvas = Canvas(bitmap)
```
I don't think anything needs to change here but I just want to note it: it's a little weird that we create a second Canvas (in the Adaptive icon case). But [the docs say](https://developer.android.com/reference/android/graphics/Canvas.html#Canvas(android.graphics.Bitmap)), "Construct a canvas with the specified bitmap to draw into." so it sounds like all the state is in the Bitmap, not the canvas, and this would be fine. It's a one time method so I don't think we care about the performance implications either.

mcomella commented on app/src/main/java/org/mozilla/focus/shortcut/IconGenerator.kt:
```diff
@@ -40,60 +33,36 @@ class IconGenerator {
+            val canvas = Canvas(bitmap)
+
+            // Draw background to comply with Adaptive Icon guidelines
+            canvas.drawColor(res.getColor(R.color.add_to_homescreen_icon_background))
```
nit: `getColor` is deprecated: you can use `ContextCompat.getColor`.
